### ISSUE TITLE
D2H copy ignores `element_arrangement` and silently returns PERMUTED data

### ISSUE THREAD

**Issue by moriohara** [torch-spyre/torch-spyre#4393]
## Summary

A tensor whose `SpyreTensorLayout.element_arrangement` is staggered
(`DL16_TO_FP32` / `FP32_TO_DL16`) reads back to the host as if it were
`STANDARD`. The host receives a **permutation** of the correct values, with **no
warning and no error**.

The values on device are correct. The EA is correct. Only the readback is wrong.

All line numbers below are against `main` @ `a62c6164`.

## Reproducer

```python
import torch
import torch_spyre  # noqa: F401

x16 = torch.ceil(torch.randn(4, 64) * 5).to(torch.float16)
xs = x16.to("spyre")

y = torch.compile(lambda a: a.float(), backend="inductor")(xs)

got = y.cpu()
ref = x16.float()
print("mismatched:", int((got != ref).sum()), "/", ref.numel())
print("same multiset:", torch.equal(torch.sort(got.flatten()).values,
                                   torch.sort(ref.flatten()).values))
```

```text
mismatched: 207 / 256
same multiset: True
```

`same multiset: True` is the diagnosis: nothing is miscomputed, the elements are
in the wrong places.

## The value is fine — only the readback is broken

The staggered EA is stamped on the live tensor:

```text
>>> y.device_tensor_layout()
SpyreTensorLayout(device_size=[2, 4, 32], stride_map=[32, 64, 1],
                  device_dtype=DataFormats.IEEE_FP32,
                  element_arrangement=ElementArrangement.DL16_TO_FP32)
```

and a **following** compiled graph consumes it correctly:

| consumer of the same staggered `y` | result |
|---|---|
| `y.cpu()` — straight to host | **PERMUTED 207/256** |
| `torch.compile(lambda b: b.half())(y)` | exact, 0/256 |
| `torch.compile(lambda b: (b * 2).half())(y)` | exact, 0/256 |
| `torch.compile(lambda b: b.sum(-1).half())(y)` | exact, 0/4 |

The reduction row is decisive: a `sum` over the **staggered dimension** would be
badly wrong if the consuming graph were not honouring the EA. The consumer's own
output reports `element_arrangement=STANDARD`, confirming the reverse conversion
restores the ordering, as the `STAGGERED_EAS` comment says it should
(`torch_spyre/_inductor/constants.py:41-46`: "can be restored by the reverse
conversion").

## How it is reached

Any stick-reordering typecast (`dl16tofp32` / `fp32todl16`, see
`_STICK_REORDERING_OPS` at `torch_spyre/_inductor/dtype_ops.py:153`) whose result
reaches a host readback. It is **not** tied to any particular op:

| `fp16 (4,64)` vs | plain `add` / `mul` | decided by |
|---|---|---|
| `(4,1)` fp32 — broadcaster **at stick dim** | **WRONG 207/256** | allowed (case 3.1, `propagate_layouts.py:1381-1388`) |
| 0-dim fp32 | **WRONG 207/256** | allowed (case 3.1, scalar exempt at `:1350-1351`) |
| `(1,64)` fp32 — broadcaster at row dim | clean `Unsupported` | rejected (case 3.3, `:1408-1431`) |
| `(4,64)` fp32 — no broadcast | clean `Unsupported` | rejected (case 3.3) |

`propagate_layouts.py` is **correct** here. `_multi_arg_pointwise_layouts`
(`:1278`) branches on `staggered_inputs = input_eas & STAGGERED_EAS` (`:1324`),
and the case analysis is spelled out in the comment at `:1302-1323`:

| case | condition | output EA |
|---|---|---|
| 1 | all STANDARD | `STANDARD` (`:1432-1434`) |
| 2a | `> 1` distinct staggered EA | `Unsupported` — unrepresentable (`:1326-1330`) |
| 3.1 | one staggered EA, every STANDARD operand can broadcast | **that staggered EA**, inherited (`:1385`) |
| 3.2 | one staggered EA, every *staggered* operand can broadcast | `STANDARD` (`:1403`) |
| 3.3 | a STANDARD and a staggered full operand coexist | `Unsupported` (`:1426-1431`) |

"Can broadcast" is `stride_map[-1] == -1` on at least one *candidate* layout
(`_broadcast_split`, `:1348-1356`) — a device-layout test, not a shape test, with
0-dim operands always broadcastable (`:1350-1351`). In the 3.1/3.2 overlap the
code prefers whichever direction holds for every candidate, defaulting to 3.1.

EA *compatibility* is deliberately not enforced at this join point — that lives in
`is_ea_compatible` (`constants.py:61`, called from `split_multi_ops.py:789`),
which is skipped here so as not to over-reject layernorm ops carrying `EXX2`
(comment at `:1302-1309`).

None of the four rows above goes through 3.2: in each, the staggered operand is
the full `(4,64)` conversion result, which has no broadcast candidate, so
`stag_can_broadcast` is false and the outcome is 3.1 or 3.3 exactly as measured.

Because case 3.1's output EA is **inherited rather than resolved**, the stagger is
contagious: every downstream pointwise op re-enters the same branch and so
demands a stick-dim broadcaster from whatever it is combined with. On a bool mask
from `fp16 (4,64) <= fp32 (4,1)`, `torch.where(mask, a, 0)` with `a` = fp16
`(4,64)` is refused, while `mask.half() * a` is exact because the conversion
restores STANDARD first. That is all self-consistent; it is only the *readback*
that ignores the EA.

For a bool result, note the physical format and the EA are resolved
independently — the format from `infer_bool_device_dtype` (`:188`, called at
`:1447-1449`, since `get_device_dtype(torch.bool)` cannot express it), the EA
from the branch above — so the mask lands as `(IEEE_FP32, DL16_TO_FP32)`, reads
back permuted 98/256, and is restored to `(SEN169_FP16, STANDARD)` exactly by
`mask.half()`.

A bare `x.float()` returned directly does the same thing. This is the mechanism
behind the known workaround that device numerics must be checked as
`.cpu().float()` and never `.float().cpu()`.

## Proposed fix

In the D2H path — `spyre__copy_from` at `torch_spyre/ops/eager.py:622`. Its
host/device branch (`:638-642`) calls `copy_tensor(self, dst, non_blocking)` at
`:641` without consulting the EA at all. Restickify there, or apply the reverse
conversion, when `self.device_tensor_layout().element_arrangement` is not
`STANDARD`. The layout is already reachable from Python at that point
(`_monkey_patch.py:119` installs the accessor), so no new plumbing is needed.

## What is *not* a fix

A compile-time pass rejecting a staggered EA on a graph output. It looks
attractive — op-agnostic, turns silent corruption into a clean error — but it
**over-rejects**: it refuses all three working cross-graph cases in the table
above. Nothing in the compiler restores the EA explicitly today, precisely
because it does not need restoring to cross a graph boundary. Refusing a
staggered graph output would remove a capability that works.

Likewise, guarding inside an individual lowering is the wrong layer: the set of
affected ops is open-ended, and by the time a `broadcast=True` pointwise lowering
runs, Inductor has already expanded both operands, so `get_size()` is identical
for the legal and illegal cases and the lowering cannot tell them apart.

## Existing coverage

None on `main` today. Coverage is proposed as part of the #3802 review:
`test_cmp_mixed_dtype_*_fp16_fp32_bcast_stick` and `..._bcast_0dim` in
`tests/inductor/test_inductor_ops.py` are `expect_fail` on a **numeric mismatch**
rather than a compile error, deliberately pinning this corruption, plus a `TODO`
at the `spyre__copy_from` site pointing here. They are `strict=True`, so they
XPASS and force their own removal once the readback honours the EA. See
[`pr3802-base...pr3802-proposed`](https://github.com/moriohara/torch-spyre/compare/pr3802-base...pr3802-proposed).

## Environment

- Code citations above are against `main` @ `a62c6164`.
- Measurements were taken on the #3802 head (`e376f010`, merge-base `d458b286`).
  The D2H site is unchanged between the two, so the corruption carries over
  verbatim. The multi-arg EA join point *has* changed on `main` — case 3.2 is new
  — but as noted above none of the measured cases reaches it.
- Single Spyre card, `SENCORES` default, `TORCHINDUCTOR_FORCE_DISABLE_CACHES=1`.

---

Related: #4392 — a different defect in the same D2H path (a zero-sized device
dim from the same class of layout rescale), which faults instead of corrupting.

**Comment by avery-blanchard** [comment]
#2255
